packages/core/src/common/performance/stopwatch.ts

```diff
@@ -0,0 +1,157 @@
+     *
+     * @param name the {@link Measurement.name name of the measurement} to wrap around the function
+     * @param description a description of what the function does, to be included in the log
+     * @param computation a supplier of the asynchronous function to wrap
+     * @param options optional addition configuration as for {@link measure}
+     * @returns the wrapped `computation`
+     *
+     * @see {@link MeasurementOptions.thresholdMillis}
+     */
+    public async startAsync<T>(name: string, description: string, computation: () => MaybePromise<T>, options?: MeasurementOptions): Promise<T> {
+        const threshold = options?.thresholdMillis ?? Number.POSITIVE_INFINITY;
+
+        const measure = this.start(name, options);
+        const result = await computation();
+        if (measure.stop() > threshold) {
+            measure.warn(`${description} took longer than the expected maximum ${threshold} milliseconds`);
+        } else {
+            measure.log(description);
+        }
+        return result;
+    }
+
+    protected createMeasurement(name: string, measurement: () => number, options?: MeasurementOptions): Measurement {
+        const logOptions = this.mergeLogOptions(options);
+
+        const result: Measurement = {
+            name,
+            stop: () => {
+                if (result.elapsed === undefined) {
+                    result.elapsed = measurement();
+                }
+                return result.elapsed;
+            },
+            log: (activity: string, ...optionalArgs: any[]) => this.log(result, activity, this.atLevel(logOptions, undefined, optionalArgs)),
+            debug: (activity: string, ...optionalArgs: any[]) => this.log(result, activity, this.atLevel(logOptions, LogLevel.DEBUG, optionalArgs)),
+            info: (activity: string, ...optionalArgs: any[]) => this.log(result, activity, this.atLevel(logOptions, LogLevel.INFO, optionalArgs)),
+            warn: (activity: string, ...optionalArgs: any[]) => this.log(result, activity, this.atLevel(logOptions, LogLevel.WARN, optionalArgs)),
+            error: (activity: string, ...optionalArgs: any[]) => this.log(result, activity, this.atLevel(logOptions, LogLevel.ERROR, optionalArgs)),
+        };
+
+        return result;
+    }
+
+    protected mergeLogOptions(logOptions?: Partial<LogOptions>): LogOptions {
+        const result: LogOptions = { ...this.defaultLogOptions };
+        if (logOptions) {
+            Object.assign(result, logOptions);
+        }
+        return result;
+    }
+
+    protected atLevel(logOptions: LogOptions, levelOverride?: LogLevel, optionalArgs?: any[]): LogOptions {
+        return { ...logOptions, levelOverride, arguments: optionalArgs };
+    }
+
+    protected logLevel(elapsed: number, options?: Partial<LogOptions>): LogLevel {
+        if (options?.levelOverride) {
+            return options?.levelOverride;
+        }
+
+        return options?.defaultLogLevel || DEFAULT_LOG_LEVEL;
```

Comment:
```suggestion
        return options?.defaultLogLevel ?? this.defaultOptions.defaultLogLevel ?? DEFAULT_LOG_LEVEL;
```
This method is likely expected to be called on options have gone through `mergeOptions`, but it isn't guaranteed, so it's like worth inserting the local default.